src/backend/access/transam/xlog.c

```diff
@@ -8811,29 +8865,106 @@ GetInsertRecPtr(void)
 }
 
 /*
- * GetLastWrittenPageLSN -- Returns maximal LSN of written page
+ * GetLastWrittenLSN -- Returns maximal LSN of written page.
+ * It returns either cached last written LSN of particular relation,
+ * either global maximum of last written LSNs among all relations.
  */
 XLogRecPtr
-GetLastWrittenPageLSN(void)
+GetLastWrittenLSN(Oid rnode, BlockNumber blkno)
 {
 	XLogRecPtr lsn;
-	SpinLockAcquire(&XLogCtl->info_lck);
-	lsn = XLogCtl->lastWrittenPageLSN;
-	SpinLockRelease(&XLogCtl->info_lck);
+	LastWrittenLsnCacheEntry* entry;
+
+	LWLockAcquire(LastWrittenLsnLock, LW_SHARED);
+
+	/* Maximal last written LSN among all non-cached pages */
+	lsn = XLogCtl->maxLastWrittenLsn;
+
+	if (rnode != InvalidOid)
+	{
+		LastWrittenLsnCacheKey key;
+		key.relid = rnode;
+		key.basket = blkno / LAST_WRITTEN_LSN_CACHE_BASKET;
+		entry = hash_search(lastWrittenLsnCache, &key, HASH_FIND, NULL);
+		if (entry != NULL)
+			lsn = entry->lsn;
+	}
+	else
+	{
+		HASH_SEQ_STATUS seq;
+		/* Find maximum of all cached LSNs */
+		hash_seq_init(&seq, lastWrittenLsnCache);
+		while ((entry = (LastWrittenLsnCacheEntry *) hash_seq_search(&seq)) != NULL)
+		{
+			if (entry->lsn > lsn)
+				lsn = entry->lsn;
+		}
+	}
```

Comment:
Scanning the hash table seems expensive. How about tracking the max of all cached LSNs a separate variable, like `maxLastWrittenLsn`?

Reply:
Sorry, but `GetLastWrittenLsn`  is called with relnode=InvalidOid only by `zenith_dbsize`. It seems to be not very frequently used  and performance critic al request so I decided to not pay extra coast for mainaining maximum.
Especially it is not so clear how to adjust it when victim is evicted from the cache.

Reply:
I do not see contradiction here. `GetLastWrittenLsn(InvaidOid)` calculates maximum among cached LSNs and `maxLastWrittenLsn` (maximal LSN for all non cached relations/buckets). So it actually returns what was originally stored in `lastWrittenLsn`.
`GetLastWrittenLsn(InvaidOid)`  is used now only by zenith_dbsize.

`SetLastWrittenLsn` i called by  createdb and dbase_redo. It updates only `maxLastWrittenLsn` . It can be that `maxLastWrittenLsn`  is less (or greater) than LSNs stored in the cache. It is ok, as far as we i any case take maximum among all of them.